### ISSUE TITLE
Parallel work?

### ISSUE THREAD

**Issue by novoxudonoser** [segmentio/nightmare#965]
I want to make a rest service what will do nightmare actions on rest call. Is it ok to make new nightmare instanse with `nightmare = Nightmare();` on rest call? Do I need to end it somehow, or `nightmare.end()` will be ok?

Is there any way to have for example pull of 10 redy to render pages and use them while requests are coming?

**Comment by rosshinkley** [comment]
There are multiple questions here, I think.

> Can Nightmare do multiple requests in parallel?

No.  Think of a Nightmare instance as a single tab on a web browser - it wouldn't make much sense to ask a tab to go to multiple locations at once, for example.

>  Is it ok to make new nightmare instanse with nightmare = Nightmare(); on rest call?

Sure.

>  Do I need to end it somehow, or nightmare.end() will be ok?

Yes, you'll need to end the Nightmare instance.  I doubt you'll want to reuse Nightmare instances across requests.

The one thing you'll need to keep in mind for the last two of the above answers: load.  Nightmare is a _bit_ of a resource hog.  If you expect your load to be fairly low and the Nightmare script is fairly quick, a new instance per request is probably reasonable.  Otherwise, you'll need to put together some sort of resource management.

> Is there any way to have for example pull of 10 redy to render pages and use them while requests are coming?

Not exactly sure what you're asking here - are you asking if Nightmare could prerender pages for a request?  Can you expand on this a bit?

**Comment by novoxudonoser** [comment]
> Not exactly sure what you're asking here - are you asking if Nightmare could prerender pages for a request? Can you expand on this a bit?

Is it a good practice to reuse nightmare ? First time I create Instance `nightmare = Nightmare();` and then I just use it for some work, after done I don't end it with `nightmare.end()` , I wait for new task, and then again. Will this be ok? Any problems like memory leaks?

**Comment by LeMoussel** [comment]
@novoxudonoser See Multiple processes: windows open and hang #1021 & @louisscruz Project [hub-crawl](https://github.com/louisscruz/hub-crawl)

**Comment by rosshinkley** [comment]
> Is it a good practice to reuse nightmare ?

You can, but be mindful of ending the instance and sharing session/partition data.  The way you have it set up, though, you're not reusing the same instance but creating a new instance every time, which is also fine.

> Any problems like memory leaks?

Great question, and the answer is probably.  It largely depends on how careful the site you're hitting is about cleaning up resources in client-side javascript.

I'm going to go ahead and close this issue for now.  Feel free to open a new issue if you have more questions.
